plugins/content/responsiveimages/responsiveimages.php

```diff
@@ -0,0 +1,171 @@
+<?php
+/**
+ * @package     Joomla.Plugin
+ * @subpackage  Content.responsiveimages
+ *
+ * @copyright   (C) 2010 Open Source Matters, Inc. <https://www.joomla.org>
+ * @license     GNU General Public License version 2 or later; see LICENSE.txt
+ */
+
+defined('_JEXEC') or die;
+
+use Joomla\CMS\Plugin\CMSPlugin;
+use Joomla\CMS\Helper\MediaHelper;
+use Joomla\CMS\Table\Table;
+
+/**
+ * Plugin for implementing responsive images with art direction functionality
+ *
+ * @since  4.1.0
+ */
+class PlgContentResponsiveImages extends CMSPlugin
+{
+	/**
+	 * Custom image sizes
+	 *
+	 * @var    string
+	 * @since  4.1.0
+	 */
+	protected $customSizes;
+
+	/**
+	 * Event that stores initial versions of images and inserts srcset and sizes
+	 * attributes into content img tags
+	 *
+	 * @param   string   $context  The context
+	 * @param   object   $table    The item
+	 * @param   boolean  $isNew    Is new item
+	 * @param   array    $data     The validated data
+	 *
+	 * @return  boolean
+	 *
+	 * @since   4.1.0
+	 */
+	public function onContentBeforeSave($context, $table, $isNew, $data)
+	{
+		// Check type of table object
+		if ($table instanceof Table)
+		{
+			$this->customSizes = MediaHelper::getSizes($this->params->get('custom_sizes'), $this->params->get('custom_size_options'));
+			$contentKey        = $this->_getContentKey($context);
+
+			// Add srcset attribute to content images
+			$table->{$contentKey} = MediaHelper::addContentSrcsetAndSizes($table->{$contentKey}, $this->customSizes);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Event that generates different sized versions of content and form images
+	 *
+	 * @param   string   $context  The context of the content passed to the plugin (added in 1.6)
+	 * @param   object   $article  A JTableContent object
+	 * @param   boolean  $isNew    If the content is just about to be created
+	 *
+	 * @return  void
+	 *
+	 * @since   4.1.0
+	 */
+	public function onContentAfterSave($context, $article, $isNew): void
+	{
+		// Check type of article object
+		if ($article instanceof Table)
+		{
+			// Generate responsive images for form and content
+			if ($formImages = $this->_getFormImages($context, (array) $article))
+			{
+				MediaHelper::generateFormResponsiveImages($formImages);
+			}
+
+			if ($content = $article->{$this->_getContentKey($context)})
+			{
+				MediaHelper::generateContentResponsiveImages($content, $this->customSizes);
+			}
+		}
+	}
+
+	/**
+	 * Returns form images from data with specific context
+	 *
+	 * @param   string  $context  The context for the data
+	 * @param   array   $data     The validated data
+	 *
+	 * @return  mixed   Array of form images or false if they don't exist
+	 *
+	 * @since   4.1.0
+	 */
+	private function _getFormImages($context, $data)
+	{
+		// Convert string data to array
+		$data['images'] = (array) json_decode($data['images']);
+		$data['params'] = (array) json_decode($data['params']);
+
+		// Get form images depending on context
+		switch ($context)
```

Comment:
It' so sad that you need to do a switch like that just because every content in joomla has different names for the same thing :/ Really good that you have the overview about all of them!